ThreeBotPackages/threefold/tfgrid_workloads/docs/reservation.md

```diff
@@ -0,0 +1,94 @@
+reservation is created (i.e. the customer signs the data). Modifications afterwards
+will cause the customer signature to become invalid.
+
+## Reservation state
+
+The reservation state is updated throughout the lifetime of the reservation. It also
+contains the signatures needed to have the farmer threebot take action. In order for
+the farmer threebot to start provisioning the workloads, or delete the workloads,
+the `signing_request_provision` and `signing_request_delete`, respectively, need to be filled with
+valid signatures. A valid signature is a signature for the reservation data, with a
+private key owned by one of the threebots listed in the reservation data (in the
+`signatures_provision` and `signatures_delete` fields). These fields also define
+the minimum amount of signatures required. For example, a signature request for
+provisioning might list 3 threebot ids which can sign, but only specify a `quorum_min`
+of 2. As such, only 2 out of the 3 listed threebot ids would need to sign before the
+farmer threebot is allowed to deploy the reserved workloads.
+
+### signature validity
+
+A signature is created by signing a piece of data using a private key. Afterwards,
+the corresponding public key can be used to check if the signature is valid. A
+[signature field](#signature.signing) is valid if it meets the following conditions:
+
+- It contains at least the minimum amount of signatures required, as defined in
+the `quorum_min` field of the corresponding [signing request field](#signing.request), 1 if there is no such
+corresponding signing request.
+- All signatures are valid with a public key owned by a referenced threebot (referenced
+in the aforementioned accompanying [signing request field](#singing.request) or possibly other field).
+
+## models
+
+The following is an overview of the models available in the [models](./models) subdirectory,
+their fields, and what these fields are used for.
+
+### reservation
+
+The [reservation object](./models/reservation.toml) is the high level object for dealing with a reservation
+on the threefold grid. It is composed of the [reservation.data](#reservation.data) object, which holds
+all data for the workloads covered by this reservation (and is immutable after it is created),
+and additional info for the reservation state.
+
+- data: As explained, the data field holds the [reservation.data](#reservation.data) object, which contains all
+low level workloads to be provisioned. After the reservation is created, this field is
+immutable.
+- json: A representation of the `data` object, in json form. It is directly derived from
+the `data` field. As such, it is also immutable.
+
+Next to the reservation data, there is also a reservation state. These fields describe
+the current state of the reservation, as well as the signatures provided by authorized threebots
+to advance the state of the reservation.
+
+- next_action: This field describes what action should be performed next. Given the enum
```

Comment:
The define flow here is the following:
created -> sign -> pay -> deploy -> delete

I feel wrong to me that there is not a more fair system in place for the payment. In this flow if the farmer node fail to deploy the workload for any reason. The farm still has already received the money. While the user doesn't have what he paid for. 
It feels to me that having a consensus from both the farmer and the user that the workload is indeed deployed is mandatory before we can unlock the money. Having this locking phase also allow the user to be automatically refunder if the workload failed to deploy.
So I would see something like
created -> sign -> TFT lock -> deploy -> farmer sign to confirm deployment -> user sign to confirm he can access -> money released

Reply:
This is likely going to be the hardest part of the entire system. Gradual release is tricky to do without a trusted third party. For this to work, we need some kind of way to periodically identify if a workload is still available, observed by a trusted party (ideally by the chain actually to avoid a 3rd party). Also, I personally don't think this always makes sense. E.g. if I deploy a workload which does a lot of computing, and outputs a single result at the end (say after a day - 24hours), it doesnt matter that it ran for 23 hours, as I don't have a result either way and I might not have any intermediate progress.

Reply:
Perhaps this discussion should be opened in a ticket so we don't forget about it. Indeed, the payment seems too optimistic here.

Reply:
lets keep things simple for now please, we are months behind and it starts to burn, we do easy for now, this is just the datamodel, if tft get on wallet of farmer he can say yes I sign, then we know enough for now